File: frontend/src/editor/profileGroups.ts
/**
 * Schritt-Gruppen fuer Bearbeitungs-Profile. Single Source ist
 * backend/schemas/edit-groups.json (das Backend liest dieselbe Datei zur
 * Laufzeit). Hier wird sie beim Build importiert — der Vite-Build-Context
 * ist das Repo-Root, daher ist der Cross-Pfad-Import zulaessig (gleiches
 * Muster wie lensProfile.ts mit infra/lensfun/profiles.json).
 */
import type { ImageEditState } from "../api/client";
import groupsData from "../../../backend/schemas/edit-groups.json";

export interface EditGroup {
  readonly key: string;
  readonly label: string;
  readonly fields: ReadonlyArray<string>;
  readonly defaultEnabled: boolean;
}

export const GROUPS = groupsData as ReadonlyArray<EditGroup>;

export function defaultEnabledGroups(): Set<string> {
  return new Set(GROUPS.filter((g) => g.defaultEnabled).map((g) => g.key));
}

const TOPLEVEL_GEOMETRY = new Set([
  "crop",
  "straightenAngle",
  "lensCorrection",
  "lensProfileId",
  "manualLensOverride",
]);

/**
 * Baut einen neuen Edit-State: startet vom `base` des Bildes und
 * ueberschreibt fuer jede angehakte Gruppe deren Felder aus `profile`.
 * Nicht-angehakte Gruppen bleiben unveraendert. Reine Funktion.
 */
export function mergeGroups(
  base: ImageEditState,
  profile: ImageEditState,
  enabled: ReadonlySet<string>,
): ImageEditState {
  const adjustments = { ...base.adjustments };
  const next: ImageEditState = {
    adjustments,
    masks: base.masks,
    crop: base.crop,
    straightenAngle: base.straightenAngle,
    lensCorrection: base.lensCorrection,
    lensProfileId: base.lensProfileId,
    manualLensOverride: base.manualLensOverride,
  };
  for (const group of GROUPS) {
    if (!enabled.has(group.key)) continue;
    for (const field of group.fields) {
      if (field === "masks") {
        next.masks = profile.masks;
      } else if (TOPLEVEL_GEOMETRY.has(field)) {
        // Top-Level-Geometrie-Feld
        (next as unknown as Record<string, unknown>)[field] =
          (profile as unknown as Record<string, unknown>)[field];
      } else {
        // Adjustment-Skalar, hsl oder toneCurve
        (adjustments as unknown as Record<string, unknown>)[field] =
          (profile.adjustments as unknown as Record<string, unknown>)[field];
      }
    }
  }
  return next;
}

File: backend/app/profile_groups.py
"""Gruppen->Feld-Mapping (Single Source: backend/schemas/edit-groups.json)
und der nicht-destruktive Merge fuer Batch-Apply.

Das Frontend importiert dieselbe JSON beim Build. Das Backend liest sie
zur Laufzeit (die Datei liegt unter backend/schemas/, wird also vom
Docker `COPY . .` ins Image uebernommen)."""
import json
from pathlib import Path
from typing import Any

_GROUPS_PATH = Path(__file__).resolve().parents[1] / "schemas" / "edit-groups.json"
with _GROUPS_PATH.open(encoding="utf-8") as _fh:
    GROUPS: list[dict[str, Any]] = json.load(_fh)

KNOWN_GROUP_KEYS: set[str] = {g["key"] for g in GROUPS}

# Felder, die unter state["adjustments"] liegen (vs. Top-Level-State-Felder).
_TOPLEVEL_GEOMETRY_FIELDS = (
    "crop", "straightenAngle", "lensCorrection",
    "lensProfileId", "manualLensOverride",
)
_DEFAULT_GEOMETRY: dict[str, Any] = {
    "crop": None,
    "straightenAngle": 0.0,
    "lensCorrection": None,
    "lensProfileId": None,
    "manualLensOverride": False,
}


def merge_edit_state(
    *,
    base_state: dict,
    preset_adjustments: dict,
    preset_masks: list,
    preset_geometry: dict | None,
    enabled: list[str],
) -> dict:
    """Baut einen neuen Edit-State (camelCase-Dict): startet vom
    `base_state` des Bildes und ueberschreibt fuer jede angehakte Gruppe
    deren Felder aus dem Preset. Nicht-angehakte Gruppen bleiben unberuehrt.

    Das Ergebnis ist ein roher Dict — der Aufrufer validiert ihn ueber
    `ImageEditState.model_validate(...)` (Ranges, Mask-Caps)."""
    enabled_set = set(enabled)
    state: dict[str, Any] = {
        "adjustments": dict(base_state.get("adjustments", {})),
        "masks": base_state.get("masks", []),
        "crop": base_state.get("crop"),
        "straightenAngle": base_state.get("straightenAngle", 0.0),
        "lensCorrection": base_state.get("lensCorrection"),
        "lensProfileId": base_state.get("lensProfileId"),
        "manualLensOverride": base_state.get("manualLensOverride", False),
    }
    geo = preset_geometry or _DEFAULT_GEOMETRY
    for group in GROUPS:
        if group["key"] not in enabled_set:
            continue
        for field in group["fields"]:
            if field == "masks":
                state["masks"] = preset_masks
            elif field in _TOPLEVEL_GEOMETRY_FIELDS:
                state[field] = geo.get(field, _DEFAULT_GEOMETRY[field])
            else:  # Adjustment-Skalar, hsl oder toneCurve
                state["adjustments"][field] = preset_adjustments.get(field)
    return state
